### PR TITLE
Add strang implicit spectral em redo

### PR DESCRIPTION
**Pull request by dpgrote**
This implements use of the PSATD field advance coupled with the implicit solver, using a Strang split advance.

- Advect Maxwell using PSATD with no J, ½ step
- Advance particles along with dE/dt = -J implicitly, full step, iterating
- Advect Maxwell using PSATD with no J, ½ step

This requires the input parameter psatd.update_with_rho = 0.
With psatd.periodic_single_box_fft = 1, exact energy conservation is obtained. Otherwise good conservation is seen, but not exact (will depend on parameters).
Convergence is found for wpedt <= 1.9 (compared to wpedt < 0.25 for FDTD).

This PR replaces PR #4662.

A task for a future PR would be to implement specialized source free spectral advance routines (as noted in source comments).

### CODE REVIEW

File: Source/FieldSolver/ImplicitSolvers/StrangImplicitSpectralEM.H [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                      bool             a_from_jacobian ) override;
+
+    [[nodiscard]] amrex::Real theta () const { return m_theta; }
+
+private:
+
+    /**
+     * \brief Time-biasing parameter for fields used on RHS to advance system
+     */
+    amrex::Real m_theta = 0.5;
```
Comment (JustinRayAngus) [inline Source/FieldSolver/ImplicitSolvers/StrangImplicitSpectralEM.H]: @dpgrote Is m_theta used?

Comment (dpgrote) [reply Source/FieldSolver/ImplicitSolvers/StrangImplicitSpectralEM.H]: I've removed references to theta.

File: Source/FieldSolver/ImplicitSolvers/StrangImplicitSpectralEM.H [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ *  The time stencil is
+ *    Advance (Eg^n, Bg^n) -> (Eg^{n+1/2}, Bg^{n+1/2}) source free // E transverse
+ *    Iterate:
+ *      Eg^{n+1} = Eg^n + c^2*dt*( - mu0*Jg^{n+1/2} ) // E longitudinal
+ *      xp^{n+1} = xp^n + dt*up^{n+1/2}/(0.5*(gammap^n + gammap^{n+1}))
+ *      up^{n+1} = up^n + dt*qp/mp*(Ep^{n+1/2} + up^{n+1/2}/gammap^{n+1/2} x Bp^{n+1/2})
+ *    Advance (Eg^n+1/2, Bg^n+1/2) -> (Eg^{n+1}, Bg^{n+1}) source free // E transverse
+ *
+ *  The algorithm is exactly energy conserving for.
+ *  The algorithm is numerially stable for any time step.
```
Comment (JustinRayAngus) [inline Source/FieldSolver/ImplicitSolvers/StrangImplicitSpectralEM.H:33]: Update comment.

File: Source/FieldSolver/ImplicitSolvers/WarpXSolverVec.H [hunk 3]
```diff
@@ -83,7 +83,8 @@ public:
     [[nodiscard]] RT dotProduct( const WarpXSolverVec&  a_X ) const;
 
     void Copy ( warpx::fields::FieldType  a_array_type,
-                warpx::fields::FieldType  a_scalar_type = warpx::fields::FieldType::None );
+                warpx::fields::FieldType  a_scalar_type = warpx::fields::FieldType::None,
+                bool allow_type_mismatch = false);
```
Comment (JustinRayAngus) [inline Source/FieldSolver/ImplicitSolvers/WarpXSolverVec.H:87]: I understand why this flag is needed here. Would be nice to not need it. That would require modifications to the WarpXSolverVec class. Ideally, m_array_type and m_scalar type would just reflect the centering for the FABs. Then, current_fp and efield_fp would be the same type. This should of course be a separate PR.

**Review by JustinRayAngus — APPROVED** [review]
Looks good Dave.
